plutus-contract/src/Wallet/Emulator/MultiAgent.hs

```diff
@@ -274,20 +283,25 @@ handleMultiAgent = interpret $ \case
         & interpret (handleZoomedState (signingProcessState wallet))
         & interpret (writeIntoState emulatorLog)
         where
-            p1 :: Prism' [EmulatorEvent] [Wallet.WalletEvent]
-            p1 = below (walletEvent wallet)
-            p2 :: Prism' [EmulatorEvent] [NC.NodeClientEvent]
-            p2 = below (walletClientEvent wallet)
-            p3 :: Prism' [EmulatorEvent] [ChainIndex.ChainIndexEvent]
-            p3 = below (chainIndexEvent wallet)
-            p4 :: Prism' [EmulatorEvent] [Log.LogMessage]
-            p4 = below (walletEvent wallet . Wallet._WalletMsg)
+            p1 :: Prism' [LogMessage EmulatorEvent] [Wallet.WalletEvent]
+            p1 = below (logMessage Info . walletEvent wallet)
+            p2 :: Prism' [LogMessage EmulatorEvent] [NC.NodeClientEvent]
+            p2 = below (logMessage Info . walletClientEvent wallet)
+            p3 :: Prism' [LogMessage EmulatorEvent] [ChainIndex.ChainIndexEvent]
+            p3 = below (logMessage Info . chainIndexEvent wallet)
+            p4 :: Prism' [LogMessage EmulatorEvent] (LogMessage T.Text)
+            p4 = _singleton . below (walletEvent wallet . Wallet._GenericLog)
+            p5 :: Prism' [LogMessage EmulatorEvent] (LogMessage RequestHandlerLogMsg)
+            p5 = _singleton . below (walletEvent wallet . Wallet._RequestHandlerLog)
+            p6 :: Prism' [LogMessage EmulatorEvent] (LogMessage TxBalanceMsg)
+            p6 = _singleton . below (walletEvent wallet . Wallet._TxBalanceLog)
```

Comment:
Y I K E S

Reply:
I have nothing specific to add otherwise.

Reply:
The only thing I can think of is maybe there could be some generic helpers for types which are sum-types where each alternative has only one argument and it's of a different type in each case. Then you could have something like

```
class Inject a b where
    inject :: a -> b

data T = A A | B B | C C

makeInject ''T -- Makes 'instance Inject A T' etc
```

Anyway, definitely not for this PR.

Reply:
https://jira.iohk.io/browse/SCP-935